cpp/src/arrow/adapters/orc/adapter_util.cc

```diff
@@ -206,15 +207,41 @@ Status AppendTimestampBatch(liborc::ColumnVectorBatch* column_vector_batch,
   const int64_t* seconds = batch->data.data() + offset;
   const int64_t* nanos = batch->nanoseconds.data() + offset;
 
+  // Firebolt Start
   auto transform_timestamp = [seconds, nanos](int64_t index) {
-    return seconds[index] * kOneSecondNanos + nanos[index];
+    // Convert ORC's timestamp column with a resolution of nanoseconds to Arrow's
+    // timestamp column with a resolution of microseconds by truncating to microseconds.
+    // We throw an exception if we detect an arithmetic overflow, as it would be hard to
+    // adapt the control flow of the existing code using Arrow's status codes. Below, we
+    // ensure that we catch this exception and translate it into a status code so that the
+    // exception does not escape the function.
+    int64_t seconds_us;
+    if (__builtin_smull_overflow(seconds[index], kOneSecondMicros, &seconds_us)) {
+      throw std::overflow_error(
+          "Overflow in ORC reader during conversion from seconds to microseconds");
+    }
+    const int64_t subseconds_us = nanos[index] / kOneMicroNanos;
+    int64_t result;
+    if (__builtin_saddl_overflow(seconds_us, subseconds_us, &result)) {
+      throw std::overflow_error(
+          "Overflow in ORC reader when adding the microseconds to the seconds");
+    }
+    return result;
   };
 
-  auto transform_range = internal::MakeLazyRange(transform_timestamp, length);
+  // Ensure that the overflow_error exception thrown in transform_timestamp cannot escape
+  // the function.
+  try {
+    auto transform_range = internal::MakeLazyRange(transform_timestamp, length);
 
-  RETURN_NOT_OK(
-      builder->AppendValues(transform_range.begin(), transform_range.end(), valid_bytes));
-  return Status::OK();
+    RETURN_NOT_OK(builder->AppendValues(transform_range.begin(), transform_range.end(),
+                                        valid_bytes));
+    return Status::OK();
+  } catch (const std::overflow_error& e) {
+    // Detected arithmetic overflow
+    return Status::Invalid(e.what());
+  }
```

Comment:
why this try catch suddenly?
why not throw let your lamda retrun an optional<string> or so which indicates if an error has occured and then return the invalid status here with the message

Reply:
If I don't use try-catch, I have to write the loop myself. The lambda is used in the MakeLazyRange and then implicitly in the builder->AppendValues and they expect the lambda to return the value and nothing else.
But writing the loop myself is also not so super easy as I need to correctly handle ORC's valid_bytes bitmap. So I would rather just use the existing code and handle the unlikely case of an overflow with an exception. Wdyt?